File: 909-snakes-and-ladders/python3/909-snakes-and-ladders.py
class Solution:
    def snakesAndLadders(self, board: List[List[int]]) -> int:
        queue = [(1, 0)]
        cost = {1: 0}
        n = len(board)
        nsq = n * n
        while len(queue) > 0:
            p, c = queue.pop(0)
            for i in range(p+1, min(p + 7, nsq+1)):
                x, y = self.p_to_x_y(i, n)
                d = i if board[y][x] == -1 else board[y][x]
                if d == nsq:
                    return c + 1
                if d in cost and cost[d] < c + 1:
                    continue
                queue.append((d, c + 1))
                cost[d] = c + 1
        return -1

    def p_to_x_y(self, p, n):
        y = n - ((p-1)//n) - 1
        x = (n*2 - (p % (n*2))) % (n*2) if (n - y) % 2 == 0 else p % (n*2) - 1
        return (x, y)

File: 909-snakes-and-ladders/java/909-snakes-and-ladders.java
class Solution {
  public int snakesAndLadders(int[][] board) {
    var n = board.length;
    var bl = new int[n * n];
    var c = 0;
    var cd = 1;
    var i = 0;
    for (var r = n - 1; r >= 0; --r) {
      while (c >=0 && c < n) {
        bl[i++] = board[r][c];
        c += cd;
      }
      cd *= -1;
      c += cd;
    }

    var dp = new int[n * n];
    for (i = 1; i < dp.length; ++i) {
      dp[i] = Integer.MAX_VALUE;
    }

    var queue = new ArrayDeque<Pair<Integer, Integer>>();
    queue.add(new Pair<>(0, 0));

    while (!queue.isEmpty()) {
      var p = queue.pollFirst();
      var square = p.getKey();
      var rolls = p.getValue();

      for (var r = 1; r <= 6; ++r) {
        var ns = square + r;
        if (ns >= (n*n)) {
          break;
        }
        if (bl[ns] != -1) {
          ns = bl[ns] - 1;
        }
        if (dp[ns] > rolls + 1) {
          dp[ns] = rolls + 1;
          queue.add(new Pair<>(ns, rolls + 1));
        }
      }
    }

    return dp[dp.length - 1] != Integer.MAX_VALUE ? dp[dp.length - 1] : -1;
  }
}

